File: src/KRATE.Core/Tools/Everyday.cs
using System.Globalization;
using System.Net;
using System.Runtime.InteropServices;

namespace Krate.Core;

public static class Everyday
{
    /// <summary>double.Parse throws a raw, untranslated FormatException that reaches the user
    /// verbatim in every language — every calculator here routes through this.</summary>
    static double[] Numbers(string input)
    {
        var parts = input.Split([' ', ',', ';', '\t', '\n', '/'], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        var values = new double[parts.Length];
        for (var i = 0; i < parts.Length; i++)
            if (!double.TryParse(parts[i], NumberStyles.Float, CultureInfo.InvariantCulture, out values[i]))
                throw new ArgumentException(Strings.Get("Error_NeedNumber"));
        return values;
    }

    static string Fmt(double v) => string.Create(CultureInfo.InvariantCulture, $"{v:0.##}");

    /// <summary>"70 175" → BMI for 70 kg and 175 cm.</summary>
    public static string Bmi(string input)
    {
        var n = Numbers(input);
        if (n.Length < 2) throw new ArgumentException(Strings.Get("Error_BmiUsage"));
        var (kg, metres) = (n[0], n[1] > 3 ? n[1] / 100 : n[1]); // accept cm or m
        if (kg <= 0 || metres <= 0) throw new ArgumentException(Strings.Get("Error_BmiUsage"));

        var bmi = kg / (metres * metres);
        var band = bmi switch
        {
            < 18.5 => "Bmi_Under",
            < 25 => "Bmi_Normal",
            < 30 => "Bmi_Over",
            _ => "Bmi_Obese",
        };
        return string.Join('\n', $"BMI  {Fmt(bmi)}", Strings.Get(band), Strings.Get("Bmi_Disclaimer"));
    }

    /// <summary>"48.50 15 3" → 15% tip on 48.50, split between 3 people.</summary>
    public static string Tip(string input)
    {
        var n = Numbers(input);
        if (n.Length == 0) throw new ArgumentException(Strings.Get("Error_NeedNumber"));
        var (bill, percent, people) = (n[0], n.Length > 1 ? n[1] : 15, n.Length > 2 ? Math.Max(1, n[2]) : 1);
        var tip = bill * percent / 100;
        var total = bill + tip;
        return string.Join('\n',
            Strings.Get("Tip_Tip", Fmt(percent), Fmt(tip)),
            Strings.Get("Tip_Total", Fmt(total)),
            people > 1 ? Strings.Get("Tip_Each", people, Fmt(total / people)) : "").TrimEnd('\n');
    }

    /// <summary>"200000 3.5 25" → monthly payment on 200 000 at 3.5% over 25 years.</summary>
    public static string Loan(string input)
    {
        var n = Numbers(input);
        if (n.Length < 3) throw new ArgumentException(Strings.Get("Error_LoanUsage"));
        var (principal, annualRate, years) = (n[0], n[1], n[2]);
        var months = (int)Math.Round(years * (years < 100 ? 12 : 1)); // years, or months if it's a big number
        var monthlyRate = annualRate / 100 / 12;

        // Standard amortisation; the zero-interest case would divide by zero.
        var payment = monthlyRate == 0
            ? principal / months
            : principal * monthlyRate / (1 - Math.Pow(1 + monthlyRate, -months));
        var total = payment * months;

        return string.Join('\n',
            Strings.Get("Loan_Monthly", Fmt(payment)),
            Strings.Get("Loan_Total", Fmt(total)),
            Strings.Get("Loan_Interest", Fmt(total - principal)),
            Strings.Get("Loan_Payments", months));
    }

    /// <summary>"192.168.1.10/24" → network, broadcast, mask, usable host range and count.</summary>
    public static string Subnet(string input)
    {
        var parts = input.Trim().Split('/');
        if (parts.Length != 2 || !IPAddress.TryParse(parts[0], out var ip) || ip.AddressFamily != System.Net.Sockets.AddressFamily.InterNetwork)
            throw new ArgumentException(Strings.Get("Error_CidrUsage"));
        var prefix = int.Parse(parts[1], CultureInfo.InvariantCulture);
        if (prefix is < 0 or > 32) throw new ArgumentException(Strings.Get("Error_CidrUsage"));

        var address = ToUInt(ip);
        var mask = prefix == 0 ? 0u : uint.MaxValue << (32 - prefix);
        var network = address & mask;
        var broadcast = network | ~mask;
        // /31 and /32 have no network/broadcast pair to exclude. Counted in 64 bits because a
        // /0 overflows a uint shift: "1u << 32" masks to "1u << 0", and 1 - 2 wrapped round to
        // 4294967295 — one host too many for the entire address space.
        var total = prefix >= 31 ? 1UL << (32 - prefix) : (1UL << (32 - prefix)) - 2;

        return string.Join('\n',
            $"NETWORK    {ToIp(network)}/{prefix}",
            $"NETMASK    {ToIp(mask)}",
            $"WILDCARD   {ToIp(~mask)}",
            $"BROADCAST  {ToIp(broadcast)}",
            $"HOSTS      {ToIp(prefix >= 31 ? network : network + 1)} – {ToIp(prefix >= 31 ? broadcast : broadcast - 1)}",
            $"USABLE     {total}",
            Strings.Get(IsPrivate(network) ? "Cidr_Private" : "Cidr_Public"));
    }

    /// <summary>Machine facts: OS, CPU, memory and disks. Input is ignored.</summary>
    public static string SysInfo(string _)
    {
        var gc = GC.GetGCMemoryInfo();
        var lines = new List<string>
        {
            $"OS         {RuntimeInformation.OSDescription}",
            $"ARCH       {RuntimeInformation.OSArchitecture}",
            $"RUNTIME    .NET {Environment.Version}",
            $"MACHINE    {Environment.MachineName}",
            $"CPU CORES  {Environment.ProcessorCount}",
            $"MEMORY     {Bytes(gc.TotalAvailableMemoryBytes)}",
        };
        foreach (var d in DriveInfo.GetDrives().Where(d => d.IsReady))
            lines.Add($"DISK {d.Name,-6} {Bytes(d.AvailableFreeSpace)} free / {Bytes(d.TotalSize)}");
        return string.Join('\n', lines);
    }

    static string Bytes(long b)
    {
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        double v = b;
        var i = 0;
        while (v >= 1024 && i < units.Length - 1) { v /= 1024; i++; }
        return string.Create(CultureInfo.InvariantCulture, $"{v:0.#} {units[i]}");
    }

    static uint ToUInt(IPAddress ip) => BitConverter.ToUInt32(ip.GetAddressBytes().Reverse().ToArray());
    static string ToIp(uint v) => new IPAddress(BitConverter.GetBytes(v).Reverse().ToArray()).ToString();

    static bool IsPrivate(uint network) =>
        (network & 0xFF000000) == 0x0A000000 ||          // 10.0.0.0/8
        (network & 0xFFF00000) == 0xAC100000 ||          // 172.16.0.0/12
        (network & 0xFFFF0000) == 0xC0A80000;            // 192.168.0.0/16

    /// <summary>GUI uses WeatherPage; CLI handles it directly. This placeholder keeps it in the searchable catalog.</summary>
    public static string Weather(string _) => throw new NotSupportedException();
    
    /// <summary>GUI uses SnakePage; CLI handles it directly. This placeholder keeps it in the searchable catalog.</summary>
    public static string Snake(string _) => throw new NotSupportedException();
    
    /// <summary>GUI uses Game2048Page; CLI handles it directly. This placeholder keeps it in the searchable catalog.</summary>
    public static string Game2048(string _) => throw new NotSupportedException();
    
    /// <summary>GUI uses TetrisPage; CLI handles it directly. This placeholder keeps it in the searchable catalog.</summary>
    public static string Tetris(string _) => throw new NotSupportedException();
}

File: rust/src/everyday.rs
//! Everyday calculators. Mirrors `Krate.Core.Everyday`.

use crate::i18n;

/// Splits on the separators the C# side accepts — note it also splits on `/`, unlike the maths
/// helper, so "70/175" works.
fn numbers(input: &str) -> Result<Vec<f64>, String> {
    input
        .split([' ', ',', ';', '\t', '\n', '/'])
        .map(str::trim)
        .filter(|p| !p.is_empty())
        .map(|p| p.parse::<f64>().map_err(|_| i18n::get("Error_NeedNumber").to_string()))
        .collect()
}

/// C#'s "0.##": at most two decimals, trailing zeros dropped. Uses the shared decimal
/// formatter, because `{:.2}` rounds the binary value and .NET rounds the shortest decimal.
fn fmt(v: f64) -> String {
    crate::tools::format_decimal(v, 2)
}

/// "70 175" gives the BMI for 70 kg and 175 cm.
pub fn bmi(input: &str) -> Result<String, String> {
    let n = numbers(input)?;
    let usage = || i18n::get("Error_BmiUsage").to_string();
    if n.len() < 2 {
        return Err(usage());
    }
    let kg = n[0];
    // Accept centimetres or metres: anything over 3 must be centimetres.
    let metres = if n[1] > 3.0 { n[1] / 100.0 } else { n[1] };
    if kg <= 0.0 || metres <= 0.0 {
        return Err(usage());
    }

    let bmi = kg / (metres * metres);
    let band = if bmi < 18.5 {
        "Bmi_Under"
    } else if bmi < 25.0 {
        "Bmi_Normal"
    } else if bmi < 30.0 {
        "Bmi_Over"
    } else {
        "Bmi_Obese"
    };
    Ok([
        format!("BMI  {}", fmt(bmi)),
        i18n::get(band).to_string(),
        i18n::get("Bmi_Disclaimer").to_string(),
    ]
    .join("\n"))
}

/// "48.50 15 3" is a 15% tip on 48.50, split between 3 people.
pub fn tip(input: &str) -> Result<String, String> {
    let n = numbers(input)?;
    if n.is_empty() {
        return Err(i18n::get("Error_NeedNumber").to_string());
    }
    let bill = n[0];
    let percent = if n.len() > 1 { n[1] } else { 15.0 };
    let people = if n.len() > 2 { n[2].max(1.0) } else { 1.0 };

    let tip = bill * percent / 100.0;
    let total = bill + tip;
    let mut lines = vec![
        i18n::format("Tip_Tip", &[&fmt(percent), &fmt(tip)]),
        i18n::format("Tip_Total", &[&fmt(total)]),
    ];
    if people > 1.0 {
        lines.push(i18n::format("Tip_Each", &[&fmt(people), &fmt(total / people)]));
    }
    Ok(lines.join("\n"))
}

/// "200000 3.5 25" is the monthly payment on 200 000 at 3.5% over 25 years.
pub fn loan(input: &str) -> Result<String, String> {
    let n = numbers(input)?;
    if n.len() < 3 {
        return Err(i18n::get("Error_LoanUsage").to_string());
    }
    let (principal, annual_rate, years) = (n[0], n[1], n[2]);
    // Years, or months already if the number is large.
    let months = (years * if years < 100.0 { 12.0 } else { 1.0 }).round() as i64;
    let monthly_rate = annual_rate / 100.0 / 12.0;

    // Standard amortisation; the zero-interest case would divide by zero.
    let payment = if monthly_rate == 0.0 {
        principal / months as f64
    } else {
        principal * monthly_rate / (1.0 - (1.0 + monthly_rate).powi(-(months as i32)))
    };
    let total = payment * months as f64;

    Ok([
        i18n::format("Loan_Monthly", &[&fmt(payment)]),
        i18n::format("Loan_Total", &[&fmt(total)]),
        i18n::format("Loan_Interest", &[&fmt(total - principal)]),
        i18n::format("Loan_Payments", &[&months.to_string()]),
    ]
    .join("\n"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bmi_accepts_centimetres_or_metres_and_bands_the_result() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        // 70 kg at 1.75 m is 22.86 — normal.
        let a = bmi("70 175").unwrap();
        let b = bmi("70 1.75").unwrap();
        assert_eq!(a, b, "cm and m must agree");
        assert!(a.contains("BMI  22.86"), "{a}");
        assert!(bmi("45 175").unwrap().lines().count() == 3);
        assert!(bmi("70").is_err(), "needs both numbers");
        assert!(bmi("0 175").is_err(), "zero weight");
        assert!(bmi("zzz").is_err());
    }

    #[test]
    fn tip_defaults_to_fifteen_percent_and_splits_only_when_asked() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        let plain = tip("100").unwrap();
        assert!(plain.contains("15"), "{plain}");
        assert_eq!(plain.lines().count(), 2, "no per-person line for one diner");
        let split = tip("100 20 4").unwrap();
        assert_eq!(split.lines().count(), 3);
        assert!(split.contains("120"), "total: {split}");
        assert!(tip("").is_err());
    }

    #[test]
    fn loan_amortises_and_survives_zero_interest() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        let m = loan("200000 3.5 25").unwrap();
        assert!(m.contains("1001.25"), "known payment for this loan: {m}");
        // Zero interest is principal split evenly, not a division by zero.
        let free = loan("12000 0 1").unwrap();
        assert!(free.contains("1000"), "{free}");
        assert!(loan("200000 3.5").is_err(), "needs three numbers");
    }

    /// .NET rounds the shortest decimal, not the binary value: a 15% tip on 48.50 is stored as
    /// 55.774999..., shown as 55.775, and rounds to 55.78 — not the 55.77 that `{:.2}` gives.
    #[test]
    fn totals_round_the_way_dotnet_prints_them() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        let split = tip("48.50 15 3").unwrap();
        assert!(split.contains("55.78"), "{split}");
        assert_eq!(fmt(55.775), "55.78");
        assert_eq!(fmt(2.5), "2.5");
        assert_eq!(fmt(2.0), "2");
        assert_eq!(fmt(-0.001), "0", "rounds to zero without a stray minus");
        assert_eq!(fmt(9.999), "10", "carry propagates into the integer part");
    }
}
